p2p/transport/webrtc/udpmux/muxed_connection.go

```diff
@@ -0,0 +1,205 @@
+)
+
+const maxPacketsInQueue = 128
+
+// muxedConnection provides a net.PacketConn abstraction
+// over packetQueue and adds the ability to store addresses
+// from which this connection (indexed by ufrag) received
+// data.
+type muxedConnection struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	pq      *packetQueue
+	ufrag   string
+	address *string
+	mux     *udpMux
+}
+
+func newMuxedConnection(mux *udpMux, ufrag string) *muxedConnection {
+	ctx, cancel := context.WithCancel(mux.ctx)
+	return &muxedConnection{
+		ctx:    ctx,
+		cancel: cancel,
+		pq:     newPacketQueue(),
+		ufrag:  ufrag,
+		mux:    mux,
+	}
+}
+
+func (conn *muxedConnection) GetAddress() (string, bool) {
+	if conn.address == nil {
+		return "", false
+	}
+	return *conn.address, true
+}
+
+func (conn *muxedConnection) SetAddress(s string) {
+	if conn.address != nil && *conn.address != s {
+		log.Debugf("address changed for ufrag %s: %s -> %s", conn.ufrag, *conn.address, s)
+	}
+	conn.address = &s
+}
+
+func (conn *muxedConnection) Push(buf []byte, addr net.Addr) error {
+	return conn.pq.Push(buf, addr)
+}
+
+// Close implements net.PacketConn
+func (conn *muxedConnection) Close() error {
+	_ = conn.closeConnection()
+	conn.mux.RemoveConnByUfrag(conn.ufrag)
+	return nil
+}
+
+// LocalAddr implements net.PacketConn
+func (conn *muxedConnection) LocalAddr() net.Addr {
+	return conn.mux.socket.LocalAddr()
+}
+
+// ReadFrom implements net.PacketConn
+func (conn *muxedConnection) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	return conn.pq.Pop(conn.ctx, p)
+}
+
+// SetDeadline implements net.PacketConn
+func (*muxedConnection) SetDeadline(t time.Time) error {
+	// no deadline is desired here
+	return nil
+}
+
+// SetReadDeadline implements net.PacketConn
+func (*muxedConnection) SetReadDeadline(t time.Time) error {
+	// no read deadline is desired here
+	return nil
+}
+
+// SetWriteDeadline implements net.PacketConn
+func (*muxedConnection) SetWriteDeadline(t time.Time) error {
+	// no write deadline is desired here
+	return nil
+}
+
+// WriteTo implements net.PacketConn
+func (conn *muxedConnection) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	return conn.mux.writeTo(p, addr)
+}
+
+func (conn *muxedConnection) closeConnection() error {
+	select {
+	case <-conn.ctx.Done():
+		return errAlreadyClosed
+	default:
+	}
+	conn.pq.Close()
+	conn.cancel()
+	return nil
+}
+
+type packet struct {
+	addr net.Addr
+	buf  []byte
+}
+
+var (
+	errTooManyPackets    = errors.New("too many packets in queue; dropping")
+	errPacketQueueClosed = errors.New("packet queue closed")
+)
+
+// just a convenience wrapper around a channel
+type packetQueue struct {
+	pktsMux sync.Mutex
+	pktsCh  chan struct{}
+	pkts    []packet
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newPacketQueue() *packetQueue {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &packetQueue{
+		pktsCh: make(chan struct{}, maxPacketsInQueue),
```

Comment:
If it's not as big as its capacity you will block when pushing multiple values at once while not yet popping. But now it will only block. as soon as you're actually empty or full, which is what you want (AFAIK).